### PR TITLE
[SPARK-36730][SQL] Use V2 Filter in V2 file source

### PR DESCRIPTION
**Pull request by huaxingao**
Co-Authored-By: DB Tsai [d_tsai@apple.com](mailto:d_tsai@apple.com)
Co-Authored-By: Huaxin Gao [huaxin_gao@apple.com](mailto:huaxin_gao@apple.com)

### What changes were proposed in this pull request?
updated the V2 file source to use V2 Filters. ParquetFilter hasn't been updated to use the V2 Filters yet and will be changed in the next PR.

### Why are the changes needed?
V2 Filter migration

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
New and existing test suites

### CODE REVIEW

**Comment by huaxingao** [comment]
cc @cloud-fan @viirya @beliefer

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVScan.scala [hunk 1]
```diff
@@ -70,8 +70,8 @@ case class CSVScan(
     ExprUtils.verifyColumnNameOfCorruptRecord(dataSchema, parsedOptions.columnNameOfCorruptRecord)
     // Don't push any filter which refers to the "virtual" column which cannot present in the input.
     // Such filters will be applied later on the upper layer.
-    val actualFilters =
-      pushedFilters.filterNot(_.references.contains(parsedOptions.columnNameOfCorruptRecord))
+    val actualFilters = pushedFilters.map(_.toV1)
+      .filterNot(_.references.contains(parsedOptions.columnNameOfCorruptRecord))
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVScan.scala:74]: Hmm, why we cannot work on v2 here and other places directly? I feel it is verbose and redundant to see `toV1`, `toV2` there.

Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVScan.scala:74]: Currently `OrcFilters`, `ParquetFilters`, `JacksonParser`, `UnivocityParser` only take v1 filters. There are actually quite some work to refactor these to make them also work with v2 filters. I prefer to have separate PRs later on for these changes.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/filter/Predicate.java [hunk 2]
```diff
@@ -146,4 +149,210 @@ public class Predicate extends GeneralScalarExpression {
...
+    String expressionStr = "";
+    for (Expression e : children()) {
+      expressionStr += e.describe() + ", ";
+    }
+
+    if (name().equals("=")) {
+      if (children()[1] instanceof LiteralValue) {
+        // e.g. a = 1
+        return new EqualTo(children()[0].describe(),
+            CatalystTypeConverters.convertToScala(((LiteralValue)children()[1]).value(),
```
Comment (beliefer) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/filter/Predicate.java:163]: Two indents

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/filter/Predicate.java [hunk 3]
```diff
@@ -146,4 +149,210 @@ public class Predicate extends GeneralScalarExpression {
...
+                ((LiteralValue)children()[1]).dataType()));
+      } else if (children()[0] instanceof LiteralValue) {
+        // e.g. 1 = a
+        return new EqualTo(children()[1].describe(),
+            CatalystTypeConverters.convertToScala(((LiteralValue)children()[0]).value(),
```
Comment (beliefer) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/filter/Predicate.java:168]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVScan.scala (hunk 1 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVScan.scala:74]: +1

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVScan.scala:74]: I see. Maybe we can create some JIRAs for these planed changes and put the JIRA number into some comments.

**Review by beliefer — COMMENTED** [review]
Uh, I just want to know why we not refactor `FileScanBuilder` by replaces `SupportsPushDownCatalystFilters` to `SupportsPushDownV2Filters` ?

File: connector/avro/src/main/scala/org/apache/spark/sql/v2/avro/AvroScan.scala [hunk 4]
```diff
@@ -59,7 +59,7 @@ case class AvroScan(
       readDataSchema,
       readPartitionSchema,
       parsedOptions,
-      pushedFilters)
+      pushedFilters.map(_.toV1))
```
Comment (beliefer) [inline connector/avro/src/main/scala/org/apache/spark/sql/v2/avro/AvroScan.scala:62]: It seems we no need `toV1` in future ?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/filter/Predicate.java [hunk 5]
```diff
@@ -146,4 +149,210 @@ public class Predicate extends GeneralScalarExpression {
   public Predicate(String name, Expression[] children) {
     super(name, children);
   }
+
+  public org.apache.spark.sql.sources.Filter toV1() {
+    String expressionStr = "";
+    for (Expression e : children()) {
+      expressionStr += e.describe() + ", ";
```
Comment (LuciferYang) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/filter/Predicate.java:156]: suggest using `StringBuilder` here

File: connector/avro/src/main/scala/org/apache/spark/sql/v2/avro/AvroScan.scala (hunk 4 above)
Comment (huaxingao) [reply connector/avro/src/main/scala/org/apache/spark/sql/v2/avro/AvroScan.scala:63]: right

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVScan.scala (hunk 1 above)
Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/csv/CSVScan.scala:75]: Added comments. Thanks

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/filter/Predicate.java (hunk 2 above)
Comment (huaxingao) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/filter/Predicate.java]: I changed my ide java continuation indent to 4 for another project. Changed back to 2 space. All the indentation should be good now.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/filter/Predicate.java (hunk 5 above)
Comment (huaxingao) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/filter/Predicate.java]: Changed. Thanks

**Comment by huaxingao** [comment]
> Uh, I just want to know why we not refactor FileScanBuilder by replaces SupportsPushDownCatalystFilters to SupportsPushDownV2Filters ?

We actually intentionally want to push down catalyst Expression instead of filter in file source, because in file source we need to do partition pruning, which uses catalyst Expression.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/filter/Predicate.java [hunk 6]
```diff
@@ -146,4 +149,210 @@ public class Predicate extends GeneralScalarExpression {
   public Predicate(String name, Expression[] children) {
     super(name, children);
   }
+
+  public org.apache.spark.sql.sources.Filter toV1() {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/filter/Predicate.java:153]: We shouldn't add this public API. Can we have a private internal util function to do it?

**Pull request closed without merging**